doc/sally.man
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SALLY 1"
.TH SALLY 1 "2019-05-08" "sally 1.0.0" "User Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
sally \- A tool for embedding strings in vector spaces
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBsally\fR [\fBoptions\fR] [\fB\-c\fR \fIconfig\fR] \fIinput\fR \fIoutput\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBsally\fR is a small tool for mapping a set of strings to a set of
vectors. This mapping is referred to as embedding and allows for
applying techniques of machine learning and data mining for analysis
of string data. \fBsally\fR can be applied to several types of string data,
such as text documents, \s-1DNA\s0 sequences or log files, where it can
handle common formats such as directories, archives and text files of
string data.
.PP
The embedding of strings is carried out incrementally, where \fBsally\fR
first loads a chunk of strings from \fIinput\fR, computes the mapping to
vectors and then writes the chunk of vectors to \fIoutput\fR. The
configuration of this process, such as the input format, the embedding
setting and the output format, are specified in the file \fIconfig\fR and
additionally using command-line options.
.PP
.Vb 7
\&    .\-\-\-\-\-\-\-\-\-.                        .\-\-\-\-\-\-\-\-\-\-.
\&    |   dir   |                        |   text   |
\&    |   arc   |   \e   .\-\-\-\-\-\-\-\-\-.   /  |  libsvm  |
\&    |  lines  |   \-\-  |  Sally  |  \-\-  |  matlab  |
\&    |   ...   |   /   \*(Aq\-\-\-\-\-\-\-\-\-\*(Aq   \e  |   ...    |
\&    \*(Aq\-\-\-\-\-\-\-\-\-\*(Aq                        \*(Aq\-\-\-\-\-\-\-\-\-\-\*(Aq
\&      Input           Embedding          Output
.Ve
.PP
\&\fBsally\fR implements a standard technique for mapping strings to a vector
space that can be referred to as generalized bag-of-words model.  The
strings are characterized by a set of features, where each feature is
associated with one dimension of the vector space.  The following types of
features are supported by \fBsally\fR: bytes, tokens, n\-grams of bytes and
n\-grams of tokens.
.PP
\&\fBsally\fR proceeds by counting the occurrences of the specified
features in each string and generating a sparse vector of count
values.  Alternatively, binary or TF-IDF values can be computed and
stored in the vectors. \fBsally\fR then normalizes the vector, for
example using the L1 or L2 norm, and outputs it in a specified format,
such as plain text, LibSVM, Cluto or Matlab format.
.PP
The processing chain of \fBsally\fR makes use of sparse data structures,
such that high-dimensional vectors with millions of dimensions can be
handled efficiently.  All features are indexed by a hash function,
where the bit width of this function can be adjusted according to the
characteristics of the data.  In several settings, a hash width
between 22 and 26 bit already suffices to store millions of features
with little loss of accuracy.
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
The configuration of \fBsally\fR is specified in a configuration file.
This file is structured into the three sections \fBinput\fR, \fBfeatures\fR
and \fBoutput\fR, which define the parameters of the input format, the
feature extraction and the output format, respectively.
.PP
All parameters of the configuration can be also be specified on the
command-line. That is, if a parameter is specified in the
configuration as \fBxx = \*(L"yy\*(R"\fR, it can be alternatively supplied as a
command-line option by \fB\-\-xx \*(L"yy\*(R"\fR.
.PP
If no configuration file is provided, \fBsally\fR resorts to a default
configuration. This default configuration can be dumped using the
command-line option \fB\-D\fR.
.SS "Input formats"
.IX Subsection "Input formats"
\&\fBsally\fR supports different formats for reading data sets of strings,
which may range from plain files to directories and other structured
resources. Following is a list of supported input formats.
.IP "\fBinput = {\fR" 4
.IX Item "input = {"
.RS 4
.PD 0
.ie n .IP "\fBinput_format = ""lines"";\fR" 4
.el .IP "\fBinput_format = ``lines'';\fR" 4
.IX Item "input_format = lines;"
.PD
This parameter specifies the input format.
.RS 4
.ie n .IP "\fI""dir""\fR" 14
.el .IP "\fI``dir''\fR" 14
.IX Item "dir"
The input strings are available as binary files in a directory and the
name of the directory is given as \fIinput\fR to \fBsally\fR. The suffixes
of the files are used as labels for the extracted vectors.
.ie n .IP "\fI""arc""\fR" 14
.el .IP "\fI``arc''\fR" 14
.IX Item "arc"
The input strings are available as binary files in a compressed
archive, such as a zip or tgz archive.  The name of the archive is
given as \fIinput\fR to \fBsally\fR.  The suffixes of the files are used as
labels for the extracted vectors.
.ie n .IP "\fI""lines""\fR" 14
.el .IP "\fI``lines''\fR" 14
.IX Item "lines"
The input strings are available as lines in a text file. The name of the
file is given as \fIinput\fR to \fBsally\fR.  The lines need to be separated by
newline and may not contain the \s-1NUL\s0 character.  Labels can be extracted from
each line using a regular expression (see \fBlines_regex\fR).
.ie n .IP "\fI""fasta""\fR" 14
.el .IP "\fI``fasta''\fR" 14
.IX Item "fasta"
The input strings are available in \s-1FASTA\s0 format. The name of the file is
given as \fIinput\fR to \fBsally\fR.  Labels can be extracted from the description
of each sequence using a regular expression (see \fBfasta_regex\fR).  Comments
are allowed if they are preceded by either ';' or '>'.
.ie n .IP "\fI""stdin""\fR" 14
.el .IP "\fI``stdin''\fR" 14
.IX Item "stdin"
The input strings are read form standard input (stdin) as text lines. The
parameter \fIinput\fR is ignored.  The lines need to be separated by newline
and may not contain the \s-1NUL\s0 character.  Labels can be extracted from each
line using a regular expression (see \fBlines_regex\fR).
.RE
.RS 4
.RE
.IP "\fBchunk_size = 256;\fR" 4
.IX Item "chunk_size = 256;"
To enable an efficient processing of large data sets, \fBsally\fR
processes strings in chunks.  This parameter defines the number of
strings in one of these chunks. Depending on the lengths of the
strings, this parameter can be adjusted to balance loading and
processing of data.
.IP "\fBdecode_str = false;\fR" 4
.IX Item "decode_str = false;"
If this parameter is set to 1, \fBsally\fR automatically decodes strings that
contain URI-encoded characters.  That is, substrings of the form \f(CW%XX\fR are

CHANGES
2012-12-27  Konrad Rieck <konrad@mlsec.org>

    - updated autotool config

2012-12-18  Konrad Rieck <konrad@mlsec.org>

    - ubuntu switched to libconfig9

2012-09-28  Konrad Rieck <konrad@mlsec.org>

    - Added authors to documentation

2012-08-28  Konrad Rieck <konrad@mlsec.org>

    - added misc docs to manual

2012-08-24  chwress <chwress@github>

    - fix some typos in the code documentation

    - nicely make use of TRUE & FALSE defines

    - make fvec_equals actually work (thx konrad)

2012-08-23  Konrad Rieck <konrad@mlsec.org>

    - new switch for default config is D

2012-08-23  chwress <chwress@github>

    - introduce idf_check for being able to properly unit-test the internal
    computed idf values

    - introduce the fvec_equals(.,.) function

    - the 1-pass of the tf*idf embedding (determining the idf values) must not
    "post-process" the raw feature vector

2012-08-22  chwress <chwress@github>

    - unify the feature vector creation

    - introduce fvec_truncate

2012-08-21  Konrad Rieck <konrad@mlsec.org>

    - implemented thresholding of vectors

    - config for thresholding

    - docs for thresholding

    - minor fixes

2012-08-21  chwress <https://github.com/chwress>

    - update the eMail address of chwress

    - naming ("print_default" -> "print_defaults")

2012-08-20  Konrad Rieck <konrad@mlsec.org>

    - added timing output

2012-08-17  Konrad Rieck <konrad@mlsec.org>

    - re-enabled openmp code

    - exit if stop word file no available but requested

    - refactored delimiter setup

    - basic framework for stopwords

    - added missing header (thanks to Andreas Ziehe)

    - stop words require the definition of delimiters

    - added missing file to automake

    - added run-time eval code to master

    - added documentation for stop words

    - support for configuring stop words

2012-08-16  Konrad Rieck <konrad@mlsec.org>

    - support for reversing of strings

    - fixed wrong email in docs

    - removed global configuration file

    - support for printing the default configuration

2012-08-13  Konrad Rieck <konrad@mlsec.org>

    - added christian as author

    - added new authors

    - fixes to stdin module

2012-08-09  Alexander Bikadorov <abiku@cs.tu-berlin.de>

    - added stdin input module

2012-08-06  Konrad Rieck <konrad@mlsec.org>

    - synchronized sally.cfg and sconfig.c (thx to chwress)

2012-05-26  Konrad Rieck <konrad@mlsec.org>

    - added md README

2012-05-14  Konrad Rieck <konrad@mlsec.org>

    - minor fixes to man page

    - fixed typo

    - fixed configure

2012-05-13  Konrad Rieck <konrad@mlsec.org>

    - fixed broken md5 copy

    - fixed gzfile pointer madness

    - added note about zlib dep

    - new release

    - spell check

2012-04-23  Konrad Rieck <konrad@mlsec.org>

    - code cleanup 6

    - code cleanup 5

    - code cleanup 4

    - code cleanup 3

    - code cleanup 2

    - code cleanup 1

2012-03-01  Konrad Rieck <konrad@mlsec.org>

    - changed legend

    - some results

2012-02-21  Konrad Rieck <konrad@mlsec.org>

    - removed some brackets

    - support for signed embedding added

    - added framework for new option

2012-02-10  Konrad Rieck <konrad@mlsec.org>

    - first cleanup of lib dir

    - destroyed structure. hmm.

2012-02-05  Konrad Rieck <konrad@mlsec.org>

    - code for sorting bytes in n-grams

    - code for sorting words in n-grams

    - added config switches

    - added support for decoding strings with URI encoding

    - switched to libtoolize

2012-02-04  Konrad Rieck <konrad@mlsec.org>

    - moved decode function to util

2012-01-22  Konrad Rieck <konrad@mlsec.org>

    - sync

2011-10-19  Konrad Rieck <konrad@mlsec.org>

    - half version of positional n-grams. explicit reprepresentation at least
    incorrect

    - config for positional n-grams

2011-07-09  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added note about dimensionality

2011-07-05  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - first version of output module for cluto

2011-04-01  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added missing config file

2011-02-21  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - new release approaching

2011-01-17  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - linux fixes

2011-01-06  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added support for global configuration and command line options

2011-01-04  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - changed names of format parameters

    - fixed regex matching

    - fixed regex code

    - added support for labels in text lines

2011-01-02  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - extended example configuration

2010-12-17  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - sanitized contributions directory

    - new contrib

2010-10-08  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - missing latex templates

    - added support to generate PDF manual

2010-10-07  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - new release 0.5.2

2010-10-05  Konrad Rieck <konrad.rieck@tu-berlin.de>


src/sally.c
/*
 * Sally - A Tool for Embedding Strings in Vector Spaces
 * Copyright (C) 2010-2014 Konrad Rieck (konrad@mlsec.org);
 *                         Christian Wressnegger (christian@mlsec.org)
 * --
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 3 of the License, or (at your
 * option) any later version.  This program is distributed without any
 * warranty. See the GNU General Public License for more details.
 */

#include "config.h"
#include "common.h"
#include "sally.h"
#include "input.h"
#include "output.h"
#include "fvec.h"
#include "util.h"
#include "reduce.h"
#include "sconfig.h"

/* Global variables */
int verbose = 0;
int print_conf = 0;
config_t cfg;

/* Local variables */
static char *input = NULL;
static char *output = NULL;
static long entries = 0;

/* Option string */
#define OPTSTRING       "g:c:i:o:n:m:r:d:psBSXE:N:b:kvqVhCD"

/**
 * Array of options of getopt_long()
 */
static struct option longopts[] = {
    {"config_file", 1, NULL, 'c'},
    {"input_format", 1, NULL, 'i'},
    {"chunk_size", 1, NULL, 1000},
    {"fasta_regex", 1, NULL, 1001},
    {"lines_regex", 1, NULL, 1002},
    {"decode_str", 0, NULL, 1005},
    {"reverse_str", 0, NULL, 1007},
    {"stoptoken_file", 1, NULL, 1008},
    {"ngram_len", 1, NULL, 'n'},
    {"granularity", 1, NULL, 'g'},
    {"token_delim", 1, NULL, 'd'},
    {"ngram_pos", 0, NULL, 'p'},
    {"pos_shift", 1, NULL, 1012},       /* <- last entry */
    {"ngram_blend", 0, NULL, 'B'},
    {"ngram_sort", 0, NULL, 's'},
    {"vect_embed", 1, NULL, 'E'},
    {"vect_norm", 1, NULL, 'N'},
    {"vect_sign", 0, NULL, 'S'},
    {"thres_low", 1, NULL, 1009},
    {"thres_high", 1, NULL, 1010},
    {"hash_bits", 1, NULL, 'b'},
    {"explicit_hash", 0, NULL, 'X'},
    {"hash_file", 1, NULL, 1011},
    {"dim_reduce", 1, NULL, 'r'},
    {"dim_num", 1, NULL, 'm'},
    {"tfidf_file", 1, NULL, 1004},
    {"output_format", 1, NULL, 'o'},
    {"skip_null", 0, NULL, 'k'},
    {"verbose", 0, NULL, 'v'},
    {"print_config", 0, NULL, 'C'},
    {"print_defaults", 0, NULL, 'D'},
    {"quiet", 0, NULL, 'q'},
    {"version", 0, NULL, 'V'},
    {"help", 0, NULL, 'h'},
    {NULL, 0, NULL, 0}
};

/**
 * Prints version and copyright information to a file stream
 * @param f File pointer
 * @param p Prefix character
 * @param m Message
 * @return number of written characters
 */
int sally_version(FILE *f, char *p, char *m)
{
    return fprintf(f, "%sSally %s - %s\n", p, PACKAGE_VERSION, m);
}

/**
 * Print configuration
 * @param msg Text to add to output
 */
static void print_config(char *msg)
{
    sally_version(stdout, "# ", msg);
    config_print(&cfg);
}

/**
 * Print usage of command line tool
 */
static void print_usage(void)
{
    printf("Usage: sally [options] <input> <output>\n"
           "\nI/O options:\n"
           "  -i,  --input_format <format>   Set input format for strings.\n"
           "       --chunk_size <num>        Set chunk size for processing.\n"
           "       --decode_str              Enable URI-decoding of strings.\n"
           "       --fasta_regex <regex>     Set RE for labels in FASTA data.\n"
           "       --lines_regex <regex>     Set RE for labels in text lines.\n"
           "       --reverse_str             Reverse (flip) the input strings.\n"
           "       --stoptoken_file <file>   Provide a file with stop tokens.\n"
           "  -o,  --output_format <format>  Set output format for vectors.\n"
           "  -k,  --skip_null               Skip null vectors in output.\n"
           "\nFeature options:\n"
           "  -n,  --ngram_len <num>         Set length of n-grams.\n"
           "  -g   --granularity <type>      Set granularity: bytes, tokens.\n"
           "  -d,  --token_delim <delim>     Set delimiters of tokens in n-grams.\n"
           "  -p,  --ngram_pos               Enable positional n-grams.\n"
           "       --pos_shift <num>         Set shift of positional n-grams.\n"
           "  -B,  --ngram_blend             Enabled blended n-grams.\n"
           "  -s,  --ngram_sort              Enable sorted n-grams (n-perms).\n"
           "  -E,  --vect_embed <embed>      Set embedding mode for vectors.\n"
           "  -N,  --vect_norm <norm>        Set normalization mode for vectors.\n"
           "  -S,  --vect_sign               Enable signed embedding.\n"
           "       --thres_low <float>       Enable minimum threshold for vectors.\n"
           "       --thres_high <float>      Enable maximum threshold for vectors.\n"
           "  -b,  --hash_bits <num>         Set number of hash bits.\n"
           "  -X,  --explicit_hash           Enable explicit hash table.\n"
           "       --hash_file <file>        Set file name for explicit hash table.\n"
           "       --tfidf_file <file>       Set file name for TFIDF weighting.\n"
           "\nFilter options:\n"
           "  -r,  --dim_reduce <method>     Set method for dimension reduction.\n"
           "  -m,  --dim_num <num>           Set number of dimensions to keep.\n"
           "\nGeneric options:\n"
           "  -c,  --config_file <file>      Set configuration file.\n"
           "  -v,  --verbose                 Increase verbosity.\n"
           "  -q,  --quiet                   Be quiet during processing.\n"
           "  -C,  --print_config            Print the current configuration.\n"
           "  -D,  --print_defaults          Print the default configuration.\n"
           "  -V,  --version                 Print version and copyright.\n"
           "  -h,  --help                    Print this help screen.\n" "\n");
}

/**
 * Print version of Sally tool
 */
static void print_version(void)
{
    printf("Sally %s - A Tool for Embedding Strings in Vector Spaces\n"
           "Copyright (c) 2010-2014 Konrad Rieck (konrad@mlsec.org)\n",
           PACKAGE_VERSION);
}

/**
 * Parse command line options
 * @param argc Number of arguments
 * @param argv Argument values
 */
static void sally_parse_options(int argc, char **argv)
{
    int ch;

    optind = 0;

    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            /* Skip. See sally_load_config(). */
            break;
        case 'i':
            config_set_string(&cfg, "input.input_format", optarg);
            break;
        case 1000:
            config_set_int(&cfg, "input.chunk_size", atoi(optarg));
            break;
        case 1001:
            config_set_string(&cfg, "input.fasta_regex", optarg);
            break;
        case 1002:
            config_set_string(&cfg, "input.lines_regex", optarg);
            break;
        case 1005:
            config_set_bool(&cfg, "input.decode_str", CONFIG_TRUE);
            break;
        case 1006:	/* Old index */
        case 'S':
            config_set_bool(&cfg, "features.vect_sign", CONFIG_TRUE);
            break;
        case 1007:
            config_set_bool(&cfg, "input.reverse_str", CONFIG_TRUE);
            break;
        case 1008:
            config_set_string(&cfg, "input.stoptoken_file", optarg);
            break;
        case 1009:
            config_set_float(&cfg, "features.thres_low", atof(optarg));
            break;
        case 1010:
            config_set_float(&cfg, "features.thres_high", atof(optarg));
            break;
        case 1011:
            config_set_string(&cfg, "features.hash_file", optarg);
            break;
        case 1012:
            config_set_int(&cfg, "features.pos_shift", atoi(optarg));
            break;
        case 'r':
            config_set_string(&cfg, "filter.dim_reduce", optarg);
            break;
        case 'm':
            config_set_int(&cfg, "filter.dim_num", atoi(optarg));
            break;
        case 'n':
            config_set_int(&cfg, "features.ngram_len", atoi(optarg));
            break;
        case 'g':
            config_set_string(&cfg, "features.granularity", optarg);
            break;
        case 'd':
            config_set_string(&cfg, "features.token_delim", optarg);
            break;
        case 'p':
            config_set_bool(&cfg, "features.ngram_pos", CONFIG_TRUE);
            break;
        case 'B':
            config_set_bool(&cfg, "features.ngram_blend", CONFIG_TRUE);
            break;
        case 's':
            config_set_bool(&cfg, "features.ngram_sort", CONFIG_TRUE);
            break;
        case 'E':
            config_set_string(&cfg, "features.vect_embed", optarg);
            break;
        case 'N':
            config_set_string(&cfg, "features.vect_norm", optarg);
            break;
        case 'b':
            config_set_int(&cfg, "features.hash_bits", atoi(optarg));
            break;
        case 1003:	/* Old index */
        case 'X':
            config_set_bool(&cfg, "features.explicit_hash", CONFIG_TRUE);
            break;
        case 1004:
            config_set_string(&cfg, "features.tfidf_file", optarg);
            break;
        case 'o':
            config_set_string(&cfg, "output.output_format", optarg);
            break;
        case 'k':
            config_set_bool(&cfg, "output.skip_null", CONFIG_TRUE);
            break;
        case 'q':
            verbose = 0;
            break;
        case 'v':
            verbose++;
            break;
        case 'D':
            print_config("Default configuration");
            exit(EXIT_SUCCESS);
            break;
        case 'C':
            print_conf = 1;
            break;
        case 'V':
            print_version();
            exit(EXIT_SUCCESS);
            break;
        case 'h':
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;
        }
    }

    /* Check configuration */
    if (!config_check(&cfg)) {
        exit(EXIT_FAILURE);
    }

    /* We are through with parsing. Print the config if requested */
    if (print_conf) {
        print_config("Current configuration");
        exit(EXIT_SUCCESS);
    }

    argc -= optind;
    argv += optind;

    /* Check for input and output arguments */
    if (argc != 2) {
        print_usage();
        exit(EXIT_FAILURE);
    } else {
        input = argv[0];
        output = argv[1];
    }

    /* Check for stdin and stdout "filenames" */
    if (!strcmp(input, "-"))
        config_set_string(&cfg, "input.input_format", "stdin");
    if (!strcmp(output, "-"))
        config_set_string(&cfg, "output.output_format", "stdout");
}


/**
 * Load the configuration of Sally
 * @param argc number of arguments
 * @param argv arguments
 */
static void sally_load_config(int argc, char **argv)
{
    char *cfg_file = NULL;
    int ch, ret;
    const char *str;

    /* Check for config file in command line */
    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            cfg_file = optarg;
            break;
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;
        default:
            /* empty */
            break;
        }
    }

    /* Init and load configuration */
    config_init(&cfg);

    if (cfg_file != NULL) {
        if (config_read_file(&cfg, cfg_file) != CONFIG_TRUE)
            fatal("Could not read configuration (%s in line %d)",
                  config_error_text(&cfg), config_error_line(&cfg));

        /* Check for new granularity parameter */
        ret = config_lookup_string(&cfg, "features.granularity", &str);
        if (ret == CONFIG_FALSE)
            fatal("Your configuration is missing the new 'granularity' "
                  "parameter. Please consult the manual page and upgrade "
                  "your configuration.");

        /* Check for old ngram_delim parameter */
        ret = config_lookup_string(&cfg, "features.ngram_delim", &str);
        if (ret == CONFIG_TRUE)
            fatal("The 'ngram_delim' parameter in your configuration "
                  "is obsolete. Please consult the manual page and use "
                  "'token_delim' instead.");

        /* Check for new token_delim parameter */
        ret = config_lookup_string(&cfg, "features.token_delim", &str);
        if (ret == CONFIG_FALSE)
            fatal("Your configuration is missing the new 'token_delim' "
                  "parameter. Please consult the manual page and upgrade "
                  "your configuration.");
    }

    /* Check configuration */
    if (!config_check(&cfg)) {
        exit(EXIT_FAILURE);
    }
}


/**
 * Init the Sally tool
 * @param argc number of arguments
 * @param argv arguments
 */
static void sally_init()
{
    int ehash;
    const char *cfg_str;

    if (verbose > 1)
        config_print(&cfg);

    /* Set delimiters */
    config_lookup_string(&cfg, "features.token_delim", &cfg_str);
    if (strlen(cfg_str) > 0)
        fvec_delim_set(cfg_str);

    /* Check for TFIDF weighting */
    config_lookup_string(&cfg, "features.vect_embed", &cfg_str);
    if (!strcasecmp(cfg_str, "tfidf"))
        idf_create(input);

    /* Load stop tokens */
    config_lookup_string(&cfg, "input.stoptoken_file", &cfg_str);
    if (strlen(cfg_str) > 0)
        stoptokens_load(cfg_str);

    /* Check for feature hash table */
    config_lookup_bool(&cfg, "features.explicit_hash", &ehash);
    config_lookup_string(&cfg, "features.hash_file", &cfg_str);
    if (ehash || strlen(cfg_str) > 0) {
        info_msg(1, "Enabling feature hash table.");
        fhash_init();
    }

    /* Open input */
    config_lookup_string(&cfg, "input.input_format", &cfg_str);
    input_config(cfg_str);
    info_msg(1, "Opening '%0.40s' with input module '%s'.", input, cfg_str);
    entries = input_open(input);
    if (entries == -1)
        fatal("Could not open input source");

    /* Open output */
    config_lookup_string(&cfg, "output.output_format", &cfg_str);
    output_config(cfg_str);
    info_msg(1, "Opening '%0.40s' with output module '%s'.", output, cfg_str);
    if (!output_open(output))
        fatal("Could not open output destination");
}

/**
 * Main processing routine of Sally. This function processes chunks of
 * strings. It might be suitable for OpenMP support in a later version.
 */
static void sally_process()
{
    long read, i, j;
    cfg_int chunk;
    const char *hash_file;

    /* Check if a hash file is set */
    config_lookup_string(&cfg, "features.hash_file", &hash_file);

    /* Get chunk size */
    config_lookup_int(&cfg, "input.chunk_size", &chunk);

    /* Allocate space */
    fvec_t **fvec = malloc(sizeof(fvec_t *) * chunk);
    string_t *strs = malloc(sizeof(string_t) * chunk);

    if (!fvec || !strs)
        fatal("Could not allocate memory for embedding");

    info_msg(1, "Processing strings in chunks of %d.", chunk);

    for (i = 0, read = 0; TRUE; i += read) {
        read = input_read(strs, chunk);
        if (read == 0)
            break;

        if (read < 0)
            fatal("Failed to read strings from input '%s'", input);

        /* Generic preprocessing of input */
        input_preproc(strs, read);

#ifdef HAVE_OPENMP
#pragma omp parallel for
#endif
        for (j = 0; j < read; j++) {
            /* Feature extraction */
            fvec[j] = fvec_extract(strs[j].str, strs[j].len);
            fvec_set_label(fvec[j], strs[j].label);
            fvec_set_source(fvec[j], strs[j].src);

            /* Dimension reduction */
            dim_reduce(fvec[j]);
        }

        if (!output_write(fvec, read))
            fatal("Failed to write vectors to output '%s'", output);

        /* Free memory */
        input_free(strs, read);
        output_free(fvec, read);

        /* Reset hash if enabled but no hash file is set */
        if (fhash_enabled() && strlen(hash_file) == 0)
            fhash_reset();

        if (entries > 0)
            prog_bar(0, entries, i + read);
    }

    free(fvec);
    free(strs);
}

/**
 * Exit Sally tool. Close open files and free memory.
 */
static void sally_exit()
{
    int ehash;
    const char *cfg_str, *hash_file;

    info_msg(1, "Flushing. Closing input and output.");
    input_close();
    output_close();

    config_lookup_string(&cfg, "features.vect_embed", &cfg_str);
    if (!strcasecmp(cfg_str, "tfidf"))
        idf_destroy(input);

    config_lookup_string(&cfg, "input.stoptoken_file", &cfg_str);
    if (strlen(cfg_str) > 0)
        stoptokens_destroy();

    config_lookup_string(&cfg, "features.hash_file", &hash_file);
    if (strlen(hash_file) > 0) {
        info_msg(1, "Saving explicit hash table to '%s'.", hash_file);
        gzFile z = gzopen(hash_file, "w9");
        if (!z)
            error("Could not open hash file '%s'", hash_file);
        fhash_write(z);
        gzclose(z);
    }

    config_lookup_bool(&cfg, "features.explicit_hash", &ehash);
    if (ehash || strlen(hash_file) > 0)
        fhash_destroy();

    /* Destroy configuration */
    config_destroy(&cfg);

}

/**
 * Main function of Sally tool
 * @param argc Number of arguments
 * @param argv Argument values
 * @return exit code
 */
int main(int argc, char **argv)
{
    sally_load_config(argc, argv);
    sally_parse_options(argc, argv);

    sally_init();
    sally_process();
    sally_exit();

    return EXIT_SUCCESS;
}
